tests/loggen/CMakeLists.txt

```diff
@@ -1,11 +1,81 @@
+# ########### loggen_helper static lib ##############
+set(LOGGEN_HELPER_SOURCE
+  loggen_helper.c
+  loggen_helper.h
+  ${CMAKE_SOURCE_DIR}/lib/crypto.c
+  ${CMAKE_SOURCE_DIR}/lib/compat/openssl_support.c
+  )
+
+add_library(loggen_helper STATIC ${LOGGEN_HELPER_SOURCE})
+
+include_directories(loggen_helper PUBLIC
+  ${CORE_INCLUDE_DIRS}
+  ${CMAKE_CURRENT_SOURCE_DIR}
+  ${GLIB_INCLUDE_DIRS}
+  )
+
+set_target_properties(loggen_helper
+    PROPERTIES VERSION ${SYSLOG_NG_VERSION}
+    SOVERSION ${SYSLOG_NG_VERSION})
+
+install(TARGETS loggen_helper ARCHIVE DESTINATION ${LOGGEN_PLUGIN_INSTALL_DIR})
+
+# ########### loggen_plugin common static lib for plugins #####
+set(LOGGEN_PLUGIN_SOURCE
+  loggen_plugin.c
+  loggen_plugin.h
+  )
+
+add_library(loggen_plugin STATIC ${LOGGEN_HELPER_SOURCE})
+
+include_directories(loggen_plugin PUBLIC
+  ${CORE_INCLUDE_DIRS}
+  ${CMAKE_CURRENT_SOURCE_DIR}
+  ${GLIB_INCLUDE_DIRS}
+  )
+
+set_target_properties(loggen_plugin
+    PROPERTIES VERSION ${SYSLOG_NG_VERSION}
+    SOVERSION ${SYSLOG_NG_VERSION})
+
+install(TARGETS loggen_plugin ARCHIVE DESTINATION ${LOGGEN_PLUGIN_INSTALL_DIR})
+
+# ########### loggen binary #################
 set(LOGGEN_SOURCE
     loggen.c
-    ${CMAKE_SOURCE_DIR}/lib/crypto.c
-    ${CMAKE_SOURCE_DIR}/lib/compat/openssl_support.c)
+    loggen_helper.c
+    loggen_helper.h
+    loggen_plugin.h
+    file_reader.c
+    file_reader.h
+    logline_generator.c
+    logline_generator.h
+    )
 
 add_executable(loggen ${LOGGEN_SOURCE})
 
-include_directories(loggen ${CORE_INCLUDE_DIRS})
-target_link_libraries(loggen syslog-ng)
+target_compile_definitions(loggen PUBLIC
+  SYSLOG_NG_PATH_LOGGEN_PLUGIN_DIR="${LOGGEN_PLUGIN_INSTALL_DIR}"
+  )
+
+include_directories(loggen
+  ${CORE_INCLUDE_DIRS}
+  ${CMAKE_CURRENT_SOURCE_DIR}
+  ${GLIB_INCLUDE_DIRS}
+  )
+
+target_link_libraries(
+  loggen
+  ${GLIB_GMODULE_LIBRARIES}
+  ${GLIB_GTHREAD_LIBRARIES}
+  ${GLIB_LIBRARIES}
+  ${OPENSSL_LIBRARIES}
+  )
 
 install(TARGETS loggen RUNTIME DESTINATION bin)
+
+# ################# build plugins ############
+set(LOGGEN_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
+add_subdirectory(socket_plugin)
+add_subdirectory(ssl_plugin)
```

Comment:
Do we need a separate ssl plugin? As I see, the code of ssl_plugin and socket plugin is very similar.

Reply:
As discussed irl, the common code is related to thread management, as of current the duplication is not a serious issue. It would be nice if we could separate those out, but it is ok the way it is now..